File: lua/orchflow/orchestrator/tmux.lua
-- orchflow.orchestrator.tmux
-- Tmux integration for terminal management

local M = {}

-- Cache for tmux executable path
local tmux_path = nil

-- Setup and verify tmux is available
function M.setup()
  -- Find tmux executable
  tmux_path = vim.fn.exepath("tmux")
  if tmux_path == "" then
    error("tmux not found in PATH. Please install tmux to use orchflow.")
  end
  
  -- Verify tmux is working
  local result = M._execute("list-sessions")
  -- It's ok if no sessions exist yet
  if result.exit_code ~= 0 and not result.output:match("no server running") then
    error("tmux not responding. Error: " .. result.output)
  end
  
  return true
end

-- Execute a tmux command
function M._execute(...)
  local args = { ... }
  local cmd = { tmux_path, unpack(args) }
  
  local output = vim.fn.system(cmd)
  local exit_code = vim.v.shell_error
  
  return {
    success = exit_code == 0,
    output = output,
    exit_code = exit_code,
  }
end

-- Create a new tmux session
function M.create_session(name, start_directory)
  local args = { "new-session", "-d", "-s", name }
  
  if start_directory then
    table.insert(args, "-c")
    table.insert(args, start_directory)
  end
  
  local result = M._execute(unpack(args))
  return result.success
end

-- Create a new pane in a session/window
function M.spawn_pane(opts)
  opts = opts or {}
  
  local args = { "split-window" }
  
  -- Target session/window
  if opts.target then
    table.insert(args, "-t")
    table.insert(args, opts.target)
  end
  
  -- Direction
  if opts.vertical then
    table.insert(args, "-v")
  else
    table.insert(args, "-h")
  end
  
  -- Size percentage
  if opts.percent then
    table.insert(args, "-p")
    table.insert(args, tostring(opts.percent))
  end
  
  -- Start directory
  if opts.cwd then
    table.insert(args, "-c")
    table.insert(args, opts.cwd)
  end
  
  -- Don't make new pane active
  if not opts.focus then
    table.insert(args, "-d")
  end
  
  -- Print pane ID
  table.insert(args, "-P")
  table.insert(args, "-F")
  table.insert(args, "#{pane_id}")
  
  -- Command to run
  if opts.command then
    table.insert(args, opts.command)
  end
  
  local result = M._execute(unpack(args))
  
  if result.success then
    local pane_id = vim.trim(result.output)
    return pane_id
  end
  
  return nil
end

-- Get information about all panes
function M.list_panes()
  local result = M._execute("list-panes", "-a", "-F", 
    "#{session_name}:#{window_index}.#{pane_index}|#{pane_id}|#{pane_pid}|#{pane_dead}|#{pane_current_command}")
  
  if not result.success then
    return {}
  end
  
  local panes = {}
  for line in result.output:gmatch("[^\n]+") do
    local target, pane_id, pid, dead, command = line:match("^([^|]+)|([^|]+)|([^|]+)|([^|]+)|(.+)$")
    if target then
      table.insert(panes, {
        target = target,
        pane_id = pane_id,
        pid = tonumber(pid),
        is_dead = dead == "1",
        command = command,
      })
    end
  end
  
  return panes
end

-- Check if a pane exists and is alive
function M.is_pane_alive(pane_id)
  local panes = M.list_panes()
  for _, pane in ipairs(panes) do
    if pane.pane_id == pane_id and not pane.is_dead then
      return true
    end
  end
  return false
end

-- Send text/command to a pane
function M.send_command(pane_id, command)
  -- Use send-keys to send the command
  local result = M._execute("send-keys", "-t", pane_id, command, "Enter")
  return result.success
end

-- Focus a specific pane
function M.focus_pane(pane_id)
  local result = M._execute("select-pane", "-t", pane_id)
  return result.success
end

-- Kill a pane
function M.kill_pane(pane_id)
  local result = M._execute("kill-pane", "-t", pane_id)
  return result.success
end

-- Kill a session
function M.kill_session(session_name)
  local result = M._execute("kill-session", "-t", session_name)
  return result.success
end

-- Get or create orchflow session
function M.get_or_create_session(name)
  name = name or "orchflow"
  
  -- Check if session exists
  local result = M._execute("has-session", "-t", name)
  
  if not result.success then
    -- Create new session
    M.create_session(name, vim.fn.getcwd())
  end
  
  return name
end

-- Attach to session (for testing)
function M.attach_session(name)
  name = name or "orchflow"
  vim.fn.system({ tmux_path, "attach-session", "-t", name })
end

-- Capture pane output
function M.capture_pane(pane_id, lines)
  lines = lines or 100
  local result = M._execute("capture-pane", "-t", pane_id, "-p", "-S", "-" .. tostring(lines))
  
  if result.success then
    return result.output
  end
  
  return ""
end

-- Resize a pane
function M.resize_pane(pane_id, direction, amount)
  local resize_flag = "-U"
  if direction == "down" then
    resize_flag = "-D"
  elseif direction == "left" then
    resize_flag = "-L"
  elseif direction == "right" then
    resize_flag = "-R"
  end
  
  local result = M._execute("resize-pane", "-t", pane_id, resize_flag, tostring(amount))
  return result.success
end

-- List all sessions
function M.list_sessions()
  local result = M._execute("list-sessions", "-F", "#{session_name}|#{session_created}|#{session_windows}")
  
  if not result.success then
    return {}
  end
  
  local sessions = {}
  for line in result.output:gmatch("[^\n]+") do
    local name, created, windows = line:match("^([^|]+)|([^|]+)|(.+)$")
    if name then
      table.insert(sessions, {
        name = name,
        created = tonumber(created),
        window_count = tonumber(windows),
      })
    end
  end
  
  return sessions
end

-- Create named window
function M.new_window(session, window_name, command)
  local args = { "new-window", "-t", session, "-n", window_name }
  
  if command then
    table.insert(args, command)
  end
  
  local result = M._execute(unpack(args))
  return result.success
end

return M

File: desktop/src-tauri/src/tmux.rs
use serde::{Deserialize, Serialize};
use std::process::Command;
use std::path::PathBuf;
use dirs::home_dir;

#[derive(Debug, Serialize, Deserialize)]
pub struct TmuxSession {
    pub name: String,
    pub windows: Vec<TmuxWindow>,
    pub created: String,
    pub attached: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TmuxWindow {
    pub id: String,
    pub name: String,
    pub panes: Vec<TmuxPane>,
    pub active: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TmuxPane {
    pub id: String,
    pub index: i32,
    pub width: i32,
    pub height: i32,
    pub command: String,
    pub active: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TmuxError {
    pub message: String,
}

pub struct TmuxManager {
    socket_path: PathBuf,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TmuxSplitResult {
    pub pane_id: String,
    pub window_id: String,
    pub session_name: String,
}

impl TmuxManager {
    pub fn new() -> Self {
        let mut socket_path = home_dir().unwrap_or_else(|| PathBuf::from("."));
        socket_path.push(".orchflow");
        socket_path.push("tmux.sock");
        
        // Ensure .orchflow directory exists
        if let Some(parent) = socket_path.parent() {
            std::fs::create_dir_all(parent).ok();
        }
        
        Self { socket_path }
    }
    
    fn tmux_cmd(&self) -> Command {
        let mut cmd = Command::new("tmux");
        cmd.arg("-S").arg(&self.socket_path);
        cmd
    }
    
    pub fn ensure_server(&self) -> Result<(), TmuxError> {
        // Check if server is running
        let output = self.tmux_cmd()
            .arg("has-session")
            .output()
            .map_err(|e| TmuxError { message: format!("Failed to check tmux server: {}", e) })?;
        
        if !output.status.success() {
            // Start new server
            self.tmux_cmd()
                .arg("new-session")
                .arg("-d")
                .arg("-s")
                .arg("orchflow-main")
                .output()
                .map_err(|e| TmuxError { message: format!("Failed to start tmux server: {}", e) })?;
        }
        
        Ok(())
    }
    
    pub fn create_session(&self, name: &str) -> Result<TmuxSession, TmuxError> {
        self.ensure_server()?;
        
        // Create new session
        let output = self.tmux_cmd()
            .arg("new-session")
            .arg("-d")
            .arg("-s")
            .arg(name)
            .arg("-P")
            .arg("-F")
            .arg("#{session_name}")
            .output()
            .map_err(|e| TmuxError { message: format!("Failed to create session: {}", e) })?;
        
        if !output.status.success() {
            return Err(TmuxError { 
                message: String::from_utf8_lossy(&output.stderr).to_string() 
            });
        }
        
        Ok(TmuxSession {
            name: name.to_string(),
            windows: vec![],
            created: chrono::Utc::now().to_rfc3339(),
            attached: false,
        })
    }
    
    pub fn list_sessions(&self) -> Result<Vec<TmuxSession>, TmuxError> {
        self.ensure_server()?;
        
        let output = self.tmux_cmd()
            .arg("list-sessions")
            .arg("-F")
            .arg("#{session_name}:#{session_created}:#{session_attached}")
            .output()
            .map_err(|e| TmuxError { message: format!("Failed to list sessions: {}", e) })?;
        
        if !output.status.success() {
            return Ok(vec![]); // No sessions
        }
        
        let sessions = String::from_utf8_lossy(&output.stdout)
            .lines()
            .filter_map(|line| {
                let parts: Vec<&str> = line.split(':').collect();
                if parts.len() >= 3 {
                    Some(TmuxSession {
                        name: parts[0].to_string(),
                        windows: vec![],
                        created: parts[1].to_string(),
                        attached: parts[2] == "1",
                    })
                } else {
                    None
                }
            })
            .collect();
        
        Ok(sessions)
    }
    
    pub fn split_pane(&self, target_pane: &str, horizontal: bool, percent: Option<u8>, command: Option<&str>) -> Result<TmuxSplitResult, TmuxError> {
        self.ensure_server()?;
        
        let mut cmd = self.tmux_cmd();
        cmd.arg("split-window");
        
        // Horizontal split (-h) creates left/right panes
        // Vertical split (default) creates top/bottom panes
        if horizontal {
            cmd.arg("-h");
        }
        
        // Target pane to split
        cmd.arg("-t").arg(target_pane);
        
        // Percentage size
        if let Some(p) = percent {
            cmd.arg("-p").arg(p.to_string());
        }
        
        // Print pane info
        cmd.arg("-P")
           .arg("-F").arg("#{pane_id}:#{window_id}:#{session_name}");
        
        // Command to run
        if let Some(command) = command {
            cmd.arg(command);
        }
        
        let output = cmd.output()
            .map_err(|e| TmuxError { message: format!("Failed to split pane: {}", e) })?;
        
        if !output.status.success() {
            return Err(TmuxError { 
                message: String::from_utf8_lossy(&output.stderr).to_string() 
            });
        }
        
        let result_info = String::from_utf8_lossy(&output.stdout).trim().to_string();
        let parts: Vec<&str> = result_info.split(':').collect();
        
        if parts.len() < 3 {
            return Err(TmuxError { message: "Invalid split result format".to_string() });
        }
        
        Ok(TmuxSplitResult {
            pane_id: parts[0].to_string(),
            window_id: parts[1].to_string(),
            session_name: parts[2].to_string(),
        })
    }
    
    pub fn create_pane(&self, session: &str, command: Option<&str>) -> Result<TmuxPane, TmuxError> {
        self.ensure_server()?;
        
        let mut cmd = self.tmux_cmd();
        cmd.arg("split-window")
            .arg("-t").arg(session)
            .arg("-P")
            .arg("-F").arg("#{pane_id}:#{pane_index}:#{pane_width}:#{pane_height}");
        
        if let Some(command) = command {
            cmd.arg(command);
        }
        
        let output = cmd.output()
            .map_err(|e| TmuxError { message: format!("Failed to create pane: {}", e) })?;
        
        if !output.status.success() {
            return Err(TmuxError { 
                message: String::from_utf8_lossy(&output.stderr).to_string() 
            });
        }
        
        let pane_info = String::from_utf8_lossy(&output.stdout).trim().to_string();
        let parts: Vec<&str> = pane_info.split(':').collect();
        
        if parts.len() < 4 {
            return Err(TmuxError { message: "Invalid pane info format".to_string() });
        }
        
        Ok(TmuxPane {
            id: parts[0].to_string(),
            index: parts[1].parse().unwrap_or(0),
            width: parts[2].parse().unwrap_or(80),
            height: parts[3].parse().unwrap_or(24),
            command: command.unwrap_or("").to_string(),
            active: true,
        })
    }
    
    pub fn send_keys(&self, pane_id: &str, keys: &str) -> Result<(), TmuxError> {
        let output = self.tmux_cmd()
            .arg("send-keys")
            .arg("-t").arg(pane_id)
            .arg(keys)
            .output()
            .map_err(|e| TmuxError { message: format!("Failed to send keys: {}", e) })?;
        
        if !output.status.success() {
            return Err(TmuxError { 
                message: String::from_utf8_lossy(&output.stderr).to_string() 
            });
        }
        
        Ok(())
    }
    
    pub fn capture_pane(&self, pane_id: &str, lines: Option<i32>) -> Result<String, TmuxError> {
        let mut cmd = self.tmux_cmd();
        cmd.arg("capture-pane")
            .arg("-t").arg(pane_id)
            .arg("-p");
        
        if let Some(n) = lines {
            cmd.arg("-S").arg(format!("-{}", n));
        }
        
        let output = cmd.output()
            .map_err(|e| TmuxError { message: format!("Failed to capture pane: {}", e) })?;
        
        if !output.status.success() {
            return Err(TmuxError { 
                message: String::from_utf8_lossy(&output.stderr).to_string() 
            });
        }
        
        Ok(String::from_utf8_lossy(&output.stdout).to_string())
    }
    
    pub fn resize_pane(&self, pane_id: &str, width: i32, height: i32) -> Result<(), TmuxError> {
        // Set width
        self.tmux_cmd()
            .arg("resize-pane")
            .arg("-t").arg(pane_id)
            .arg("-x").arg(width.to_string())
            .output()
            .map_err(|e| TmuxError { message: format!("Failed to resize pane width: {}", e) })?;
        
        // Set height
        self.tmux_cmd()
            .arg("resize-pane")
            .arg("-t").arg(pane_id)
            .arg("-y").arg(height.to_string())
            .output()
            .map_err(|e| TmuxError { message: format!("Failed to resize pane height: {}", e) })?;
        
        Ok(())
    }
    
    pub fn kill_pane(&self, pane_id: &str) -> Result<(), TmuxError> {
        let output = self.tmux_cmd()
            .arg("kill-pane")
            .arg("-t").arg(pane_id)
            .output()
            .map_err(|e| TmuxError { message: format!("Failed to kill pane: {}", e) })?;
        
        if !output.status.success() {
            return Err(TmuxError { 
                message: String::from_utf8_lossy(&output.stderr).to_string() 
            });
        }
        
        Ok(())
    }
}

// Tauri commands
#[tauri::command]
pub async fn tmux_create_session(name: String) -> Result<TmuxSession, String> {
    let manager = TmuxManager::new();
    manager.create_session(&name).map_err(|e| e.message)
}

#[tauri::command]
pub async fn tmux_split_pane(
    target_pane: String, 
    horizontal: bool, 
    percent: Option<u8>, 
    command: Option<String>
) -> Result<TmuxSplitResult, String> {
    let manager = TmuxManager::new();
    manager.split_pane(&target_pane, horizontal, percent, command.as_deref())
        .map_err(|e| e.message)
}

#[tauri::command]
pub async fn tmux_list_sessions() -> Result<Vec<TmuxSession>, String> {
    let manager = TmuxManager::new();
    manager.list_sessions().map_err(|e| e.message)
}

#[tauri::command]
pub async fn tmux_create_pane(session: String, command: Option<String>) -> Result<TmuxPane, String> {
    let manager = TmuxManager::new();
    manager.create_pane(&session, command.as_deref()).map_err(|e| e.message)
}

#[tauri::command]
pub async fn tmux_send_keys(pane_id: String, keys: String) -> Result<(), String> {
    let manager = TmuxManager::new();
    manager.send_keys(&pane_id, &keys).map_err(|e| e.message)
}

#[tauri::command]
pub async fn tmux_capture_pane(pane_id: String, lines: Option<i32>) -> Result<String, String> {
    let manager = TmuxManager::new();
    manager.capture_pane(&pane_id, lines).map_err(|e| e.message)
}

#[tauri::command]
pub async fn tmux_resize_pane(pane_id: String, width: i32, height: i32) -> Result<(), String> {
    let manager = TmuxManager::new();
    manager.resize_pane(&pane_id, width, height).map_err(|e| e.message)
}

#[tauri::command]
pub async fn tmux_kill_pane(pane_id: String) -> Result<(), String> {
    let manager = TmuxManager::new();
    manager.kill_pane(&pane_id).map_err(|e| e.message)
}